Ark.Tools.NLog/Slack/SlackMessageBuilder.cs

```diff
@@ -0,0 +1,108 @@
+﻿using Slack.Webhooks;
+
+using System;
+using System.Collections.Generic;
+
+namespace Ark.Tools.NLog.Slack
+{
+    public class SlackMessageBuilder
+    {
+        private readonly string _webHookUrl;
+        private readonly SlackMessage _slackMessage;
+        private event Action<Exception> Error;
+        private bool _existError = false;
+
+        public SlackMessageBuilder(string webHookUrl)
+        {
+            this._webHookUrl = webHookUrl;
+            _slackMessage = new SlackMessage();
+        }
+
+        public static SlackMessageBuilder Build(string webHookUrl)
+        {
+            return new SlackMessageBuilder(webHookUrl);
+        }
+
+        public SlackMessageBuilder WithMessage(string message)
+        {
+            this._slackMessage.Text = message;
+
+            return this;
+        }
+
+        public SlackMessageBuilder AddAttachment(string color, IEnumerable<(string title, string text)> fields)
+        {
+            var a = new SlackAttachment
+            {
+                Color = color
+            };
+
+            a.Fields = new List<SlackField>();
+
+            foreach (var (title, text) in fields)
+                a.Fields.Add(new SlackField()
+                {
+                    Title = title,
+                    Value = text,
+                    Short = true
+                });
+
+            if (this._slackMessage.Attachments == null)
+                this._slackMessage.Attachments = new List<SlackAttachment>();
+
+            this._slackMessage.Attachments.Add(a);
+
+            return this;
+        }
+
+        public SlackMessageBuilder AddAttachment(string message, string color, IEnumerable<(string title, string text)> fields)
+        {
+            var a = new SlackAttachment
+            {
+                Text = message,
+                Color = color
+            };
+
+            a.Fields = new List<SlackField>();
+
+            foreach (var (title, text) in fields)
+                a.Fields.Add(new SlackField()
+                {
+                    Title = title,
+                    Value = text,
+                    Short = true
+                });
+
+            if (this._slackMessage.Attachments == null)
+                this._slackMessage.Attachments = new List<SlackAttachment>();
+
+            this._slackMessage.Attachments.Add(a);
+
+            return this;
+        }
+
+        public SlackMessageBuilder OnError(Action<Exception> error)
+        {
+            this.Error += error;
+            _existError = true;
+
+            return this;
+        }
+
+        public void Send()
+        {
+            try
+            {
+                using var c = new SlackClient(this._webHookUrl);
```

Comment:
SlackClient is an HttpClient wrapper. HttpClient is expensive to create and should be cached or IHttpClientFactory should be used which cache the inner Handler (the expensive part).

When used for many messages, it count as each HttpClient would re-estabilish TLS with Slack if not cached and exhaush port due to TCP lifetime :(

https://learn.microsoft.com/en-us/dotnet/core/extensions/httpclient-factory
https://www.aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/

SlackClient could be parte of the SlackTarget, created on Init and Disposed with the Target.